crypto/ffc/ffc_params_generate.c

```diff
@@ -0,0 +1,992 @@
+            goto err;
+        }
+        /* A.2.4 Step (13): Return valid if computed_g == g */
+        if (validate_flags && BN_cmp(g, params->g) != 0) {
+            *res = FFC_CHECK_G_MISMATCH;
+            goto err;
+        }
+    } else if (generate) {
+        if (!generate_unverifiable_g(ctx, mont, g, tmp, p, e, pm1, &hret))
+            goto err;
+    }
+
+    if (!BN_GENCB_call(cb, 3, 1))
+        goto err;
+
+    if (generate) {
+        if (p != params->p) {
+            BN_free(params->p);
+            params->p = BN_dup(p);
+        }
+        if (q != params->q) {
+            BN_free(params->q);
+            params->q = BN_dup(q);
+        }
+        if (g != params->g) {
+            BN_free(params->g);
+            params->g = BN_dup(g);
+        }
+        if (params->p == NULL || params->q == NULL || params->g == NULL)
+            goto err;
+        if (!ffc_params_set_validate_params(params, seed, seedlen, pcounter))
+            goto err;
+        params->h = hret;
+    }
+pass:
+    if ((validate_flags & FFC_PARAMS_VALIDATE_G) != 0 && (canonical_g == 0))
+        /* Return for the case where g is partially valid */
+        ok = FFC_PARAMS_RET_STATUS_UNVERIFIABLE_G;
+    else
+        ok = FFC_PARAMS_RET_STATUS_SUCCESS;
+err:
+    if (seed != params->seed)
+        OPENSSL_free(seed);
+    OPENSSL_free(seed_tmp);
+    if (ctx)
+        BN_CTX_end(ctx);
+    BN_CTX_free(ctx);
+    BN_MONT_CTX_free(mont);
+    EVP_MD_CTX_free(mctx);
+    return ok;
+}
+
+int ffc_param_FIPS186_2_gen_verify(OPENSSL_CTX *libctx, FFC_PARAMS *params,
+                                   int type, size_t L, size_t N,
+                                   const EVP_MD *evpmd, int validate_flags,
+                                   int *res, BN_GENCB *cb)
+{
+    int ok = FFC_PARAMS_RET_STATUS_FAILED;
+    unsigned char seed[SHA256_DIGEST_LENGTH];
+    unsigned char buf[SHA256_DIGEST_LENGTH];
+    BIGNUM *r0, *test, *tmp, *g = NULL, *q = NULL, *p = NULL;
+    BN_MONT_CTX *mont = NULL;
+    size_t qsize = N >> 3;
+    int n = 0, m = 0;
+    int counter = 0, pcounter = 0, use_random_seed;
+    int rv;
+    BN_CTX *ctx = NULL;
+    int hret = -1;
+    int generate = (validate_flags == 0);
+    unsigned char *seed_in = params->seed;
+    size_t seed_len = params->seedlen;
+
+    *res = 0;
+#ifdef FIPS_MODE
+    /*
+     * FIPS 186-4 states that validation can only be done for this pair.
+     * (Even though the original spec allowed L = 512 + 64*j (j = 0.. 8))
+     */
+    if (L != 1024 || N != 160) {
+        *res = FFC_CHECK_BAD_LN_PAIR;
+        return FFC_PARAMS_RET_STATUS_FAILED;
+    }
+#endif
+    if (qsize != SHA_DIGEST_LENGTH
+        && qsize != SHA224_DIGEST_LENGTH
+        && qsize != SHA256_DIGEST_LENGTH) {
+        /* invalid q size */
+        *res = FFC_CHECK_INVALID_Q_VALUE;
+        return FFC_PARAMS_RET_STATUS_FAILED;
+    }
+
+    if (evpmd == NULL) {
+        if (qsize == SHA_DIGEST_LENGTH)
+            evpmd = EVP_sha1();
```

Comment:
Does this even compile in FIPS_MODE? I don't think using implicitly fetched digests like this will work in the FIPS provider. (Side note: Perhaps we should prevent their use by wrapping the declarations in `#ifndef FIPS_MODE`)

Reply:
It does compile.. I would prefer to leave it in so will change it to do a fetch. (We shouldnt break existing code just because it is FIPS if we can help it).

Reply:
It's actually quite concerning that this built without an error. I just checked, and legacy_sha.c which provides the implementation for EVP_sha1() does not get linked into fips.so....so how is the link completing successfully???

Reply:
I investigated this further, and this turns out to be a real problem in our linking. You *should* have got a linker error from this. See #11020.

Reply:
That may imply that there might be a few more issues :)